### PR TITLE
Refactor internal vars installer

### LINKED ISSUE #797: Upgrading from a directory fails if the IMAGE variable exists in environment files
**cos-toolkit version:**
cOS-Seed-green-0.7.0-12-g3e2e084a-x86_64.is

**CPU architecture, OS, and Version:**

Linux cos 5.3.18-59.24-default #1 SMP Mon Sep 13 15:06:42 UTC 2021 (2f872ea) x86_64 x86_64 x86_64 GNU/Linux

**Describe the bug**

If there is an `IMAGE=xxx` line in `/etc/os-release` file, running `cos-upgrade` command to upgrade from a directory fails.

**To Reproduce**

- Add an `IMAGE=xxxx` line:
```
cos:~ # cp /etc/os-release /tmp/
cos:~ # echo >> /tmp/os-release
cos:~ # echo 'IMAGE="quay.io/costoolkit/releases-green:cos-system-0.7.0-14"' >> /tmp/os-release
cos:~ # mount -o bind /tmp/os-release /etc/os-release
cos:~ # cat /etc/os-release
ANSI_COLOR="0;32"
BUG_REPORT_URL="https://github.com/mudler/cOS/issues"
DOCUMENTATION_URL="https://github.com/mudler/cOS"
HOME_URL="https://github.com/mudler/cOS"
ID="cOS"
NAME="cOS"
PRETTY_NAME="cOS v0.7.0-12"
VERSION="0.7.0-12"
IMAGE="quay.io/costoolkit/releases-green:cos-system-0.7.0-14"
```

- Mount a squashfs image to a directory and upgrade from it
```
# I upload the squashfs from an ISO to /usr/local 
cos:~ # ls -l /usr/local/rootfs.squashfs
-rw-r--r-- 1 root root 314286080 Oct 14 08:10 /usr/local/rootfs.squashfs
cos:~ # mkdir /tmp/new
cos:~ # mount /usr/local/rootfs.squashfs /tmp/new/
cos:~ # cos-upgrade --directory /tmp/new/
Upgrading to image quay.io/costoolkit/releases-green:cos-system-0.7.0-14
Upgrading system..
-> Upgrade target: active.img
3240+0 records in
3240+0 records out
3397386240 bytes (3.4 GB, 3.2 GiB) copied, 1.54999 s, 2.2 GB/s
mke2fs 1.43.8 (1-Jan-2018)
Discarding device blocks: done
Creating filesystem with 829440 4k blocks and 207584 inodes
Filesystem UUID: 41da2a4e-4fc3-4c06-acfc-89f5756f4e92
Superblock backups stored on blocks:
	32768, 98304, 163840, 229376, 294912, 819200

Allocating group tables: done
Writing inode tables: done
Writing superblocks and filesystem accounting information: done

mkdir: cannot create directory ‘/tmp/upgrade/usr/local’: No such file or directory
INFO[0000] yip version 0.9.14-g8cc8e0e3e42aa4d70715cc5ef97693f0d490d61d 2021-10-11 05:03:41 UTC
INFO[0000] Executing /system/oem/00_rootfs.yaml
INFO[0000] Executing yip file                            name="Rootfs Layout Settings" stage=before-upgrade.before stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.before stages=0 success=true
INFO[0000] Executing /system/oem/01_defaults.yaml
INFO[0000] Executing yip file                            name="General settings" stage=before-upgrade.before stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.before stages=0 success=true
INFO[0000] Executing /system/oem/02_upgrades.yaml
INFO[0000] Executing yip file                            name="Upgrades settings" stage=before-upgrade.before stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.before stages=0 success=true
INFO[0000] Executing /system/oem/03_branding.yaml
INFO[0000] Executing yip file                            name=Branding stage=before-upgrade.before stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.before stages=0 success=true
INFO[0000] Executing /system/oem/04_accounting.yaml
INFO[0000] Executing yip file                            name="Default user" stage=before-upgrade.before stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.before stages=0 success=true
INFO[0000] Executing /system/oem/05_network.yaml
INFO[0000] Executing yip file                            name="Default network configuration" stage=before-upgrade.before stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.before stages=0 success=true
INFO[0000] Executing /system/oem/06_recovery.yaml
INFO[0000] Executing yip file                            name="Recovery partition boot setup" stage=before-upgrade.before stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.before stages=0 success=true
INFO[0000] Executing /system/oem/07_live.yaml
INFO[0000] Executing yip file                            name="LiveCD Detection" stage=before-upgrade.before stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.before stages=0 success=true
INFO[0000] yip version 0.9.14-g8cc8e0e3e42aa4d70715cc5ef97693f0d490d61d 2021-10-11 05:03:41 UTC
INFO[0000] Executing /system/oem/00_rootfs.yaml
INFO[0000] Executing yip file                            name="Rootfs Layout Settings" stage=before-upgrade stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade stages=0 success=true
INFO[0000] Executing /system/oem/01_defaults.yaml
INFO[0000] Executing yip file                            name="General settings" stage=before-upgrade stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade stages=0 success=true
INFO[0000] Executing /system/oem/02_upgrades.yaml
INFO[0000] Executing yip file                            name="Upgrades settings" stage=before-upgrade stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade stages=0 success=true
INFO[0000] Executing /system/oem/03_branding.yaml
INFO[0000] Executing yip file                            name=Branding stage=before-upgrade stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade stages=0 success=true
INFO[0000] Executing /system/oem/04_accounting.yaml
INFO[0000] Executing yip file                            name="Default user" stage=before-upgrade stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade stages=0 success=true
INFO[0000] Executing /system/oem/05_network.yaml
INFO[0000] Executing yip file                            name="Default network configuration" stage=before-upgrade stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade stages=0 success=true
INFO[0000] Executing /system/oem/06_recovery.yaml
INFO[0000] Executing yip file                            name="Recovery partition boot setup" stage=before-upgrade stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade stages=0 success=true
INFO[0000] Executing /system/oem/07_live.yaml
INFO[0000] Executing yip file                            name="LiveCD Detection" stage=before-upgrade stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade stages=0 success=true
INFO[0000] yip version 0.9.14-g8cc8e0e3e42aa4d70715cc5ef97693f0d490d61d 2021-10-11 05:03:41 UTC
INFO[0000] Executing /system/oem/00_rootfs.yaml
INFO[0000] Executing yip file                            name="Rootfs Layout Settings" stage=before-upgrade.after stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.after stages=0 success=true
INFO[0000] Executing /system/oem/01_defaults.yaml
INFO[0000] Executing yip file                            name="General settings" stage=before-upgrade.after stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.after stages=0 success=true
INFO[0000] Executing /system/oem/02_upgrades.yaml
INFO[0000] Executing yip file                            name="Upgrades settings" stage=before-upgrade.after stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.after stages=0 success=true
INFO[0000] Executing /system/oem/03_branding.yaml
INFO[0000] Executing yip file                            name=Branding stage=before-upgrade.after stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.after stages=0 success=true
INFO[0000] Executing /system/oem/04_accounting.yaml
INFO[0000] Executing yip file                            name="Default user" stage=before-upgrade.after stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.after stages=0 success=true
INFO[0000] Executing /system/oem/05_network.yaml
INFO[0000] Executing yip file                            name="Default network configuration" stage=before-upgrade.after stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.after stages=0 success=true
INFO[0000] Executing /system/oem/06_recovery.yaml
INFO[0000] Executing yip file                            name="Recovery partition boot setup" stage=before-upgrade.after stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.after stages=0 success=true
INFO[0000] Executing /system/oem/07_live.yaml
INFO[0000] Executing yip file                            name="LiveCD Detection" stage=before-upgrade.after stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.after stages=0 success=true
INFO[0000] yip version 0.9.14-g8cc8e0e3e42aa4d70715cc5ef97693f0d490d61d 2021-10-11 05:03:41 UTC
INFO[0000] yip version 0.9.14-g8cc8e0e3e42aa4d70715cc5ef97693f0d490d61d 2021-10-11 05:03:41 UTC
INFO[0000] yip version 0.9.14-g8cc8e0e3e42aa4d70715cc5ef97693f0d490d61d 2021-10-11 05:03:41 UTC
INFO[0000] yip version 0.9.14-g8cc8e0e3e42aa4d70715cc5ef97693f0d490d61d 2021-10-11 05:03:41 UTC
INFO[0000] Executing /proc/cmdline
INFO[0000] Executing yip file                            name= stage=before-upgrade.before stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.before stages=0 success=true
INFO[0000] yip version 0.9.14-g8cc8e0e3e42aa4d70715cc5ef97693f0d490d61d 2021-10-11 05:03:41 UTC
INFO[0000] Executing /proc/cmdline
INFO[0000] Executing yip file                            name= stage=before-upgrade stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade stages=0 success=true
INFO[0000] yip version 0.9.14-g8cc8e0e3e42aa4d70715cc5ef97693f0d490d61d 2021-10-11 05:03:41 UTC
INFO[0000] Executing /proc/cmdline
INFO[0000] Executing yip file                            name= stage=before-upgrade.after stages=0
INFO[0000] Finished yip file execution                   stage=before-upgrade.after stages=0 success=true
Upgrading from local folder: quay.io/costoolkit/releases-green:cos-system-0.7.0-14
rsync: change_dir "/root//quay.io/costoolkit/releases-green:cos-system-0.7.0-14" failed: No such file or directory (2)
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1189) [sender=3.1.3]
```

**Expected behavior**

The upgrade should succeed.
The upgrade source should be the specified directory rather than the IMAGE variable value.

**Logs**

**Additional context**

The [`IMAGE` variable set in the parse opts loop](https://github.com/rancher-sandbox/cOS-toolkit/blob/49eb210468fa7a70a94e1c0cf8ced3ec2d66cd11/packages/installer/upgrade.sh#L352) is overrided after sourcing some environment files: https://github.com/rancher-sandbox/cOS-toolkit/blob/49eb210468fa7a70a94e1c0cf8ced3ec2d66cd11/packages/installer/upgrade.sh#L66-L89

### LINKED ISSUE #799: Differentiate internal variables
**Is your feature request related to a problem? Please describe.**
#797 shows how easy is to interact with internal variables from config files. the `IMAGE` variable is not intended to be configured or set via config file, but if that happens the upgrade process is broken.

**Describe the solution you'd like**
Variables that are meant to be used internally inside the scripts shouldn't conflict with global variables. Use variable prefixes (e.g. `cos_image`, `_cos_image` instead of `IMAGE`) or whatever necessary to avoid clashes

**Describe alternatives you've considered**

**Additional context**
#797

### PR DESCRIPTION
**Pull request by Itxaka**
Change all vars used in the script to have a `_` in front so no collisions with existing vars
Use more locals
Set proper override of values in order Defaults -> config -> env vars -> switches
The change above means that I was able to remove some of the assignments and move them all to the same load_env_vars function

Fixes #799 
Fixes #797

### CODE REVIEW

**Comment by Itxaka** [comment]
soooo

It has 63 commits because in case that it fails its easier to bisect. It will be squashed

File: packages/installer/cos.sh
Comment (mudler) [inline packages/installer/cos.sh:13]: This one is configurable by a user. Shall we maybe prefix all of the ones that are configurable?

For backward compatibility we should also respect `GRUBCONF` in this case too.

File: packages/installer/cos.sh
Comment (mudler) [inline packages/installer/cos.sh:16]: This one is intended to be tweakable too

File: packages/installer/cos.sh
Comment (mudler) [inline packages/installer/cos.sh:574]: See here. `CURRENT` can actually be used to avoid prompting and select the target image. Maybe this one needs a prefix too.

File: packages/installer/cos.sh
Comment (mudler) [inline packages/installer/cos.sh:758]: Those can be specified via config as well. It's a way to have defaults.

A specific prefix as well here makes sense

**Review by mudler — CHANGES_REQUESTED** [review]
We have to be a bit careful here. There are variables that are meant to be configured (e.g. the one we have in our default example configuration file), and some which can be specified via CLI, e.g. `CURRENT` while selecting default targets (mostly used in scripting). 

Besides, we have to offer some backward compatibility or we are going to break user configurations default.

File: packages/installer/cos.sh
Comment (Itxaka) [reply packages/installer/cos.sh:13]: That is true.

My idea is to move the configurable ones to be set on the `load_config` and use only internal ones. So no touching any external ones or using them at all, so they only entrypoints for the script are the switches or the supported configs mentioned under https://rancher-sandbox.github.io/cos-toolkit-docs/docs/customizing/general_configuration/

File: packages/installer/cos.sh
Comment (Itxaka) [inline packages/installer/cos.sh:38]: tested this manually to check the order:

```
$ ./cos.sh upgrade
where is verify coming from: config
$ VERIFY=ENV ./cos.sh upgrade
where is verify coming from: ENV
$ VERIFY=ENV ./cos.sh upgrade --no-verify
where is verify coming from: flag
$ ./cos.sh upgrade --no-verify 
where is verify coming from: flag
```

Message output was a temp change to set the proper value on the `_VERIFY` var, print that and exit

File: packages/installer/cos.sh
Comment (Itxaka) [reply packages/installer/cos.sh:13]: should be fixed by the change below

File: packages/installer/cos.sh
Comment (Itxaka) [reply packages/installer/cos.sh:16]: should be fixed by the change below

File: packages/installer/cos.sh
Comment (mudler) [inline packages/installer/cos.sh:720]: Do we need to update the help message ?

Comment (Itxaka) [reply packages/installer/cos.sh:720]: why? am Im missing something here? Is CURRENT able to be overwritten? 'cause its not documented

File: packages/installer/cos.sh
Comment (mudler) [inline packages/installer/cos.sh:591]: mmh too bad we don't have tests for suc-upgrade. I remember we loaded config here because otherwise the config file would override what was the user provided flags from CLI, breaking it. https://github.com/rancher-sandbox/cOS-toolkit/pull/543/files#r693834604

File: packages/installer/cos.sh
Comment (mudler) [reply packages/installer/cos.sh:720]: `    # We booted from an ISO or some else medium. We assume we want to fixup the current label
        read -p "Could not determine current partition. Do you want to overwrite your current active partition? (CURRENT=active.img) [y/N] : 
`

it is few lines below if no `CURRENT` was found. Basically that is the only way to recover other images (besides active) for e.g. when booting from a LiveCD

File: packages/installer/cos.sh
Comment (Itxaka) [reply packages/installer/cos.sh:591]: isnt this then fixed with the new order load?

File: packages/installer/cos.sh
Comment (Itxaka) [reply packages/installer/cos.sh:720]: Sorry, Im still not getting the problem here, could you elaborate further?

Is current supposed to be able to overwrite via env var?

File: packages/installer/cos.sh
Comment (mudler) [reply packages/installer/cos.sh:591]: I think it is indeed, currently having a closer look :eyes:

File: packages/installer/cos.sh
Comment (mudler) [reply packages/installer/cos.sh:720]: yep, the script tries to find a "current" image as target for the upgrade process but it is a best effort, hence why the command prompt. Since the prompt could break automatic flows (scripts, etc) it is meant to be possible to set it to avoid any interaction

Comment (Itxaka) [reply packages/installer/cos.sh:720]: ah, ok. Do we support the CURRENT in the config file or only via env vars?

Comment (mudler) [reply packages/installer/cos.sh:720]: I don't think it makes sense really sense in the config file as it is a corner-case here. Besides defaults wouldn't be relevant as we try to auto-detect it

Comment (Itxaka) [reply packages/installer/cos.sh:725]: updated!

**Review by mudler — APPROVED** [review]
Looks good!

**Review by mudler — CHANGES_REQUESTED** [review]
right, it needs a bump of cOS packages or.. tests wouldn't run on this PR :)

**Comment by Itxaka** [comment]
LOOOL

**Review by mudler — APPROVED** [review]

**Comment by mudler** [comment]
Needs a tag of 0.7.4 before moving ahead and merging this

File: packages/installer/cos.sh
Comment (mudler) [inline packages/installer/cos.sh:1407]: mmh maybe this is not necessary, as it makes sense only if we have UPGRADE_IMAGE set

**Comment by mudler** [comment]
This PR is (yet) another case where https://github.com/rancher-sandbox/cOS-toolkit/issues/108 would fit in the scope instead

File: packages/installer/cos.sh
Comment (Itxaka) [reply packages/installer/cos.sh:1407]: indeed, no matter if this is set or not in this path

**Comment by Itxaka** [comment]
Maybe this should wait until the installer tests on #859 are expanded so we can make sure that we are covering the installer correctly and not breaking stuff. After all #797 has a workaround, so its not high priority.

**Comment by Itxaka** [comment]
Blocked by #859

**Comment by mudler** [comment]
Having a tagged release in between should be sufficient too, so we can always point to those artifacts if needed.. we could tackle regressions individually later

**Comment by Itxaka** [comment]
> Having a tagged release in between should be sufficient too, so we can always point to those artifacts if needed.. we could tackle regressions individually later

Sounds good but...do we need to? I rather not merge broken code, and fix in this PR any issues that I can find while running the tests. Plus, we gotta add more tests anyway so...

Also helps me figure out if I missed something in the rework and maybe fix it better (like the config stuff!)

**Comment by Itxaka** [comment]
as soon as https://github.com/rancher-sandbox/cOS-toolkit/pull/888 is merged, I will rebase this and see if it passes...

**Comment by Itxaka** [comment]
Rebased. Lets see if those installer test also pass now....

**Pull request merged**
